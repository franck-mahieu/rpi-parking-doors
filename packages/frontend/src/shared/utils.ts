const env = process.env.NODE_ENV;
const reactAppDevApiUrl = process.env.REACT_APP_DEV_API_URL;

export const getBaseUrl = (
  envParam: string | undefined,
  reactAppApiUrlParam: string | undefined,
) => {
  return envParam === 'development' ? reactAppApiUrlParam : '';
};

export const fetchUtils = async (uri: string, method: string, body?: any) => {
  const response = await fetch(`${getBaseUrl(env, reactAppDevApiUrl)}${uri}`, {
    method: method,
    headers: {
      'Content-Type': 'application/json',
    },
    credentials: env === 'development' ? 'include' : 'same-origin',
    body: body,
  });
  const result: any = await response.json();
  if (
    (response.status >= 200 && response.status < 300) ||
    (result?.changes && result.changes !== 0)
  ) {
    return result;
  }
  return Promise.reject(result);
};
