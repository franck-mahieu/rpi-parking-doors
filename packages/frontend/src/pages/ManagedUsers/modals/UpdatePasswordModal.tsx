import React, { Component } from 'react';
import M from 'materialize-css';
import 'materialize-css/dist/css/materialize.min.css';
import { fetchUtils } from '../../../shared/utils';
import './Modals.scss';

const {
  REACT_APP_UPDATE_PASSWORD_MODAL_TITLE,
  REACT_APP_UPDATE_PASSWORD_SUCCESS_MESSAGE,
  REACT_APP_UPDATE_PASSWORD_ERROR_MESSAGE,
  REACT_APP_LABEL_BUTTON_CANCEL,
  REACT_APP_LABEL_BUTTON_VALIDATE,
  REACT_APP_LABEL_PASSORD,
  REACT_APP_LABEL_UPDATE_PASSWORD_BUTTON,
} = process.env;

interface IUpdatePasswordModal {
  login: string;
  guid: string;
}

interface IUpdatePasswordModalState {
  password: string | undefined;
}

class UpdatePasswordModal extends Component<IUpdatePasswordModal> {
  private Modal: any;
  public state: IUpdatePasswordModalState;

  constructor(props: IUpdatePasswordModal) {
    super(props);
    this.state = {
      password: '',
    };
  }

  componentDidMount() {
    const onOpenStart = () => {
      this.setState({
        password: '',
      });
    };
    const options = {
      onOpenStart,
      inDuration: 250,
      outDuration: 250,
      opacity: 0.5,
      dismissible: true,
      startingTop: '4%',
      endingTop: '10%',
    };
    M.Modal.init(this.Modal, options);
  }

  onClickUpdatePassword = async (
    login: string,
    password: string | undefined,
    guid: string,
  ) => {
    try {
      await fetchUtils(
        `/api/users/updatepassword?guid=${guid}`,
        'POST',
        JSON.stringify({ login, password }),
      );
      M.toast({
        html: `${login}${REACT_APP_UPDATE_PASSWORD_SUCCESS_MESSAGE}`,
        classes: 'green',
      });
    } catch (e) {
      M.toast({
        html: `${login}${REACT_APP_UPDATE_PASSWORD_ERROR_MESSAGE}`,
        classes: 'red',
      });
    }
  };

  onChangePasswordHandler = (event: any) => {
    this.setState({
      password: event?.target?.value,
    });
  };

  render() {
    return (
      <span id={'user-modal'}>
        <button
          className="waves-effect waves-light btn modal-trigger update-password-button"
          data-target={`update-user-password-modal-${this.props.login}`}
          title={REACT_APP_LABEL_UPDATE_PASSWORD_BUTTON}
        >
          {REACT_APP_LABEL_UPDATE_PASSWORD_BUTTON}
        </button>

        <div
          ref={Modal => {
            this.Modal = Modal;
          }}
          id={`update-user-password-modal-${this.props.login}`}
          className="modal"
        >
          <div className="modal-content">
            <h4>
              {REACT_APP_UPDATE_PASSWORD_MODAL_TITLE} {this.props.login}
            </h4>
            <div className="row">
              <div className="input-field col s12">
                <input
                  id={`update-password-${this.props.login}`}
                  type="text"
                  className="validate"
                  value={this.state.password}
                  onChange={this.onChangePasswordHandler}
                />
                <label htmlFor="password">{REACT_APP_LABEL_PASSORD}</label>
              </div>
            </div>
          </div>
          <div className="modal-footer">
            <button className="modal-close waves-effect waves-red btn-flat cancel-button">
              {REACT_APP_LABEL_BUTTON_CANCEL}
            </button>
            <button
              id={'update-password-validate-button'}
              className="modal-close waves-effect waves-green btn-flat"
              onClick={async () => {
                await this.onClickUpdatePassword(
                  this.props.login,
                  this.state.password,
                  this.props.guid,
                );
              }}
            >
              {REACT_APP_LABEL_BUTTON_VALIDATE}
            </button>
          </div>
        </div>
      </span>
    );
  }
}

export default UpdatePasswordModal;
