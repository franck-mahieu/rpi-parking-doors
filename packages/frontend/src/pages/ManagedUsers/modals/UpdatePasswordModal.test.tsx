import React from 'react';
import renderer from 'react-test-renderer';
import UpdatePasswordModal from './UpdatePasswordModal';

import * as utils from '../../../shared/utils';
import M from 'materialize-css';

jest.mock('../../../shared/utils');

describe('UpdatePasswordModal', () => {
  let updatePasswordModal: UpdatePasswordModal;
  beforeEach(() => {
    jest.resetModules();
    jest.restoreAllMocks();
    //@ts-ignore
    utils.fetchUtils.mockResolvedValue(true);
    updatePasswordModal = new UpdatePasswordModal({
      login: 'login',
      guid: 'guid',
    });
  });

  describe('render', () => {
    it('renders correctly', () => {
      const tree = renderer
        .create(<UpdatePasswordModal guid={'guid'} login={'login'} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should call onClickUpdatePassword on click on up button', () => {
      const component = renderer.create(
        <UpdatePasswordModal guid={'guid'} login={'login'} />,
      );
      const componentInstance: any = component.getInstance();
      componentInstance.onChangePasswordHandler({
        target: { value: 'new password' },
      });
      jest
        .spyOn(componentInstance, 'onClickUpdatePassword')
        .mockResolvedValue('');
      component.root
        .findByProps({ id: 'update-password-validate-button' })
        .props.onClick();
      expect(componentInstance.onClickUpdatePassword).toHaveBeenCalledWith(
        'login',
        'new password',
        'guid',
      );
    });
  });

  describe('componentDidMount', () => {
    jest.mock('materialize-css');
    it('should call M.Modal.init with valid params', async () => {
      const modalInitMock = jest.fn().mockImplementation((modal, options) => {
        options.onOpenStart();
      });
      M.Modal.init = modalInitMock;
      const setStateMock = jest.fn();
      updatePasswordModal.setState = setStateMock;
      updatePasswordModal.componentDidMount();
      expect(modalInitMock).toHaveBeenCalledWith(undefined, {
        onOpenStart: expect.any(Function),
        inDuration: 250,
        outDuration: 250,
        opacity: 0.5,
        dismissible: true,
        startingTop: '4%',
        endingTop: '10%',
      });
      expect(setStateMock).toHaveBeenCalledWith({
        password: '',
      });
    });
  });

  describe('onClickUpdatePassword', () => {
    it('should call fetchUtils and call M.toast with green when fetch is in sucess', async () => {
      const toastMock = jest.fn();
      M.toast = toastMock;
      await updatePasswordModal.onClickUpdatePassword(
        'login',
        'password',
        'guid',
      );
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/users/updatepassword?guid=guid',
        'POST',
        '{"login":"login","password":"password"}',
      );
      expect(toastMock).toBeCalledWith({
        html: 'login, le mot de passe a été mis à jour avec succès',
        classes: 'green',
      });
    });
    it('should call fetchUtils and call M.toast with red when fetch is in error', async () => {
      const toastMock = jest.fn();
      M.toast = toastMock;
      //@ts-ignore
      utils.fetchUtils.mockRejectedValue();
      await updatePasswordModal.onClickUpdatePassword(
        'login',
        'password',
        'guid',
      );
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/users/updatepassword?guid=guid',
        'POST',
        '{"login":"login","password":"password"}',
      );
      expect(toastMock).toBeCalledWith({
        html: `login, le mot de passe n'a pas été mis à jour, une erreur s'est produite`,
        classes: 'red',
      });
    });
  });

  describe('onChangePasswordHandler', () => {
    it('should call setState with pasword received', () => {
      const setStateMock = jest.fn();
      updatePasswordModal.setState = setStateMock;
      updatePasswordModal.onChangePasswordHandler({
        target: { value: 'password' },
      });
      expect(setStateMock).toHaveBeenCalledWith({ password: 'password' });
    });

    it('should call setState with undefined when received empty event', () => {
      const setStateMock = jest.fn();
      updatePasswordModal.setState = setStateMock;
      updatePasswordModal.onChangePasswordHandler({});
      expect(setStateMock).toHaveBeenCalledWith({ password: undefined });
    });
  });
});
