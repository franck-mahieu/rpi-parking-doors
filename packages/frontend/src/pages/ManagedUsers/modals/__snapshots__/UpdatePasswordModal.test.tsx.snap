// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePasswordModal render renders correctly 1`] = `
<span
  id="user-modal"
>
  <button
    className="waves-effect waves-light btn modal-trigger update-password-button"
    data-target="update-user-password-modal-login"
    title="Mise à jour du mot de passe"
  >
    Mise à jour du mot de passe
  </button>
  <div
    className="modal"
    id="update-user-password-modal-login"
  >
    <div
      className="modal-content"
    >
      <h4>
        Mise à jour du mot de passe de
         
        login
      </h4>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="validate"
            id="update-password-login"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            htmlFor="update-password-login"
          >
            Mot de passe
          </label>
        </div>
      </div>
    </div>
    <div
      className="modal-footer"
    >
      <button
        className="modal-close waves-effect waves-red btn-flat cancel-button"
      >
        Annuler
      </button>
      <button
        className="modal-close waves-effect waves-green btn-flat"
        id="update-password-validate-button"
        onClick={[Function]}
      >
        Valider
      </button>
    </div>
  </div>
</span>
`;
